File: packages/python-sdk/src/schemax/storage_v3.py
"""
Storage Layer V3 - Provider-Aware

New storage layer that supports multiple catalog providers through the provider system.
Migrates from v2 storage format to v3 format with provider metadata.
"""

import json
import hashlib
from pathlib import Path
from typing import Any, Dict, List, Optional
from datetime import datetime
from uuid import uuid4

from .providers import ProviderRegistry, Provider, Operation, ProviderState

SCHEMAX_DIR = ".schemax"
PROJECT_FILENAME = "project.json"
CHANGELOG_FILENAME = "changelog.json"
SNAPSHOTS_DIR = "snapshots"


def get_schemax_dir(workspace_path: Path) -> Path:
    """Get the .schemax directory path"""
    return workspace_path / SCHEMAX_DIR


def get_project_file_path(workspace_path: Path) -> Path:
    """Get the project file path"""
    return get_schemax_dir(workspace_path) / PROJECT_FILENAME


def get_changelog_file_path(workspace_path: Path) -> Path:
    """Get the changelog file path"""
    return get_schemax_dir(workspace_path) / CHANGELOG_FILENAME


def get_snapshots_dir(workspace_path: Path) -> Path:
    """Get the snapshots directory path"""
    return get_schemax_dir(workspace_path) / SNAPSHOTS_DIR


def get_snapshot_file_path(workspace_path: Path, version: str) -> Path:
    """Get snapshot file path"""
    return get_snapshots_dir(workspace_path) / f"{version}.json"


def ensure_schemax_dir(workspace_path: Path) -> None:
    """Ensure .schemax/snapshots/ directory exists"""
    schemax_dir = get_schemax_dir(workspace_path)
    snapshots_dir = get_snapshots_dir(workspace_path)

    schemax_dir.mkdir(parents=True, exist_ok=True)
    snapshots_dir.mkdir(parents=True, exist_ok=True)


def ensure_project_file(workspace_path: Path, provider_id: str = "unity") -> None:
    """
    Initialize a new project with provider selection

    Args:
        workspace_path: Path to workspace
        provider_id: Provider ID (default: 'unity')
    """
    project_path = get_project_file_path(workspace_path)
    changelog_path = get_changelog_file_path(workspace_path)

    if project_path.exists():
        # Project exists - check if migration is needed
        with open(project_path, "r") as f:
            project = json.load(f)

        if project.get("version") == 2:
            # Migrate from v2 to v3
            migrate_v2_to_v3(workspace_path, project, provider_id)
        return

    # Create new v3 project
    workspace_name = workspace_path.name

    # Verify provider exists
    provider = ProviderRegistry.get(provider_id)
    if provider is None:
        available = ", ".join(ProviderRegistry.get_all_ids())
        raise ValueError(
            f"Provider '{provider_id}' not found. Available providers: {available}"
        )

    new_project = {
        "version": 3,
        "name": workspace_name,
        "provider": {
            "type": provider_id,
            "version": provider.info.version,
        },
        "environments": ["dev", "test", "prod"],
        "snapshots": [],
        "deployments": [],
        "settings": {
            "autoIncrementVersion": True,
            "versionPrefix": "v",
            "requireSnapshotForProd": True,
            "allowDrift": False,
            "requireComments": False,
            "warnOnBreakingChanges": True,
        },
        "latestSnapshot": None,
    }

    new_changelog = {
        "version": 1,
        "sinceSnapshot": None,
        "ops": [],
        "lastModified": datetime.utcnow().isoformat() + "Z",
    }

    ensure_schemax_dir(workspace_path)

    with open(project_path, "w") as f:
        json.dump(new_project, f, indent=2)

    with open(changelog_path, "w") as f:
        json.dump(new_changelog, f, indent=2)

    print(
        f"[SchemaX] Initialized new v3 project: {workspace_name} with provider: {provider.info.name}"
    )


def migrate_v2_to_v3(
    workspace_path: Path, v2_project: Dict[str, Any], provider_id: str = "unity"
) -> None:
    """
    Migrate v2 project to v3 format

    Args:
        workspace_path: Path to workspace
        v2_project: V2 project data
        provider_id: Provider ID (default: 'unity')
    """
    print("[SchemaX] Migrating project from v2 to v3...")

    provider = ProviderRegistry.get(provider_id)
    if provider is None:
        raise ValueError(f"Provider '{provider_id}' not found")

    v3_project = {
        "version": 3,
        "name": v2_project.get("name", workspace_path.name),
        "provider": {
            "type": provider_id,
            "version": provider.info.version,
        },
        "environments": v2_project.get("environments", ["dev", "test", "prod"]),
        "snapshots": v2_project.get("snapshots", []),
        "deployments": v2_project.get("deployments", []),
        "settings": v2_project.get(
            "settings",
            {
                "autoIncrementVersion": True,
                "versionPrefix": "v",
                "requireSnapshotForProd": True,
                "allowDrift": False,
                "requireComments": False,
                "warnOnBreakingChanges": True,
            },
        ),
        "latestSnapshot": v2_project.get("latestSnapshot"),
    }

    # Migrate operations in changelog to add provider prefix
    try:
        changelog = read_changelog(workspace_path)
        migrated_ops = []

        for op in changelog["ops"]:
            # Add provider field if missing
            if "provider" not in op:
                op["provider"] = provider_id

            # Add provider prefix to operation type if missing
            if "." not in op["op"]:
                op["op"] = f"{provider_id}.{op['op']}"

            migrated_ops.append(op)

        changelog["ops"] = migrated_ops

        with open(get_changelog_file_path(workspace_path), "w") as f:
            json.dump(changelog, f, indent=2)

    except Exception as e:
        print(f"[SchemaX] Warning: Could not migrate changelog operations: {e}")

    # Write migrated project file
    with open(get_project_file_path(workspace_path), "w") as f:
        json.dump(v3_project, f, indent=2)

    print("[SchemaX] Migration complete: v2 → v3")


def read_project(workspace_path: Path) -> Dict[str, Any]:
    """
    Read project file

    Args:
        workspace_path: Path to workspace

    Returns:
        Project data
    """
    project_path = get_project_file_path(workspace_path)

    with open(project_path, "r") as f:
        project = json.load(f)

    # Auto-migrate if needed
    if project.get("version") == 2:
        migrate_v2_to_v3(workspace_path, project)
        return read_project(workspace_path)  # Re-read after migration

    if project.get("version") != 3:
        raise ValueError(f"Unsupported project version: {project.get('version')}")

    return project


def write_project(workspace_path: Path, project: Dict[str, Any]) -> None:
    """Write project file"""
    project_path = get_project_file_path(workspace_path)

    with open(project_path, "w") as f:
        json.dump(project, f, indent=2)


def read_changelog(workspace_path: Path) -> Dict[str, Any]:
    """
    Read changelog file

    Args:
        workspace_path: Path to workspace

    Returns:
        Changelog data
    """
    changelog_path = get_changelog_file_path(workspace_path)

    try:
        with open(changelog_path, "r") as f:
            return json.load(f)
    except FileNotFoundError:
        # Changelog doesn't exist, create empty one
        changelog = {
            "version": 1,
            "sinceSnapshot": None,
            "ops": [],
            "lastModified": datetime.utcnow().isoformat() + "Z",
        }

        with open(changelog_path, "w") as f:
            json.dump(changelog, f, indent=2)

        return changelog


def write_changelog(workspace_path: Path, changelog: Dict[str, Any]) -> None:
    """Write changelog file"""
    changelog_path = get_changelog_file_path(workspace_path)
    changelog["lastModified"] = datetime.utcnow().isoformat() + "Z"

    with open(changelog_path, "w") as f:
        json.dump(changelog, f, indent=2)


def read_snapshot(workspace_path: Path, version: str) -> Dict[str, Any]:
    """Read a snapshot file"""
    snapshot_path = get_snapshot_file_path(workspace_path, version)

    with open(snapshot_path, "r") as f:
        return json.load(f)


def write_snapshot(workspace_path: Path, snapshot: Dict[str, Any]) -> None:
    """Write a snapshot file"""
    ensure_schemax_dir(workspace_path)
    snapshot_path = get_snapshot_file_path(workspace_path, snapshot["version"])

    with open(snapshot_path, "w") as f:
        json.dump(snapshot, f, indent=2)


def load_current_state(
    workspace_path: Path,
) -> tuple[ProviderState, Dict[str, Any], Provider]:
    """
    Load current state using provider

    Args:
        workspace_path: Path to workspace

    Returns:
        Tuple of (state, changelog, provider)
    """
    project = read_project(workspace_path)
    changelog = read_changelog(workspace_path)

    # Get provider
    provider = ProviderRegistry.get(project["provider"]["type"])
    if provider is None:
        raise ValueError(
            f"Provider '{project['provider']['type']}' not found. "
            "Please ensure the provider is installed."
        )

    # Load state
    if project["latestSnapshot"]:
        # Load latest snapshot
        snapshot = read_snapshot(workspace_path, project["latestSnapshot"])
        state = snapshot["state"]
    else:
        # No snapshots yet, start with empty state
        state = provider.create_initial_state()

    # Apply changelog ops using provider's state reducer
    ops = [Operation(**op) for op in changelog["ops"]]
    state = provider.apply_operations(state, ops)

    return state, changelog, provider


def append_ops(workspace_path: Path, ops: List[Operation]) -> None:
    """
    Append operations to changelog

    Args:
        workspace_path: Path to workspace
        ops: Operations to append
    """
    project = read_project(workspace_path)
    changelog = read_changelog(workspace_path)
    provider = ProviderRegistry.get(project["provider"]["type"])

    if provider is None:
        raise ValueError(f"Provider '{project['provider']['type']}' not found")

    # Validate operations
    for op in ops:
        validation = provider.validate_operation(op)
        if not validation.valid:
            errors = ", ".join(f"{e.field}: {e.message}" for e in validation.errors)
            raise ValueError(f"Invalid operation: {errors}")

    # Append ops
    changelog["ops"].extend([op.dict(by_alias=True) for op in ops])

    # Write back
    write_changelog(workspace_path, changelog)


def create_snapshot(
    workspace_path: Path,
    name: str,
    version: Optional[str] = None,
    comment: Optional[str] = None,
    tags: List[str] = None,
) -> tuple[Dict[str, Any], Dict[str, Any]]:
    """
    Create a snapshot

    Args:
        workspace_path: Path to workspace
        name: Snapshot name
        version: Optional version string
        comment: Optional comment
        tags: Optional tags

    Returns:
        Tuple of (project, snapshot)
    """
    if tags is None:
        tags = []

    project = read_project(workspace_path)
    changelog = read_changelog(workspace_path)

    # Load current state
    state, _, _ = load_current_state(workspace_path)

    # Determine version
    snapshot_version = version or _get_next_version(
        project.get("latestSnapshot"), project["settings"]
    )

    # Generate IDs for ops that don't have them (backwards compatibility)
    ops_with_ids = []
    for i, op in enumerate(changelog["ops"]):
        if "id" not in op or not op["id"]:
            op["id"] = f"op_{i}_{op['ts']}_{op['target']}"
        ops_with_ids.append(op)

    # Calculate hash
    state_hash = _calculate_state_hash(state, [op["id"] for op in ops_with_ids])

    import os

    # Create snapshot file
    snapshot_file = {
        "id": f"snap_{uuid4()}",
        "version": snapshot_version,
        "name": name,
        "ts": datetime.utcnow().isoformat() + "Z",
        "createdBy": os.environ.get("USER") or os.environ.get("USERNAME") or "unknown",
        "state": state,
        "opsIncluded": [op["id"] for op in ops_with_ids],
        "previousSnapshot": project.get("latestSnapshot"),
        "hash": state_hash,
        "tags": tags,
        "comment": comment,
    }

    # Write snapshot file
    write_snapshot(workspace_path, snapshot_file)

    # Create snapshot metadata
    snapshot_metadata = {
        "id": snapshot_file["id"],
        "version": snapshot_version,
        "name": name,
        "ts": snapshot_file["ts"],
        "createdBy": snapshot_file["createdBy"],
        "file": f".schemax/snapshots/{snapshot_version}.json",
        "previousSnapshot": project.get("latestSnapshot"),
        "opsCount": len(ops_with_ids),
        "hash": state_hash,
        "tags": tags,
        "comment": comment,
    }

    # Update project
    project["snapshots"].append(snapshot_metadata)
    project["latestSnapshot"] = snapshot_version
    write_project(workspace_path, project)

    # Clear changelog
    new_changelog = {
        "version": 1,
        "sinceSnapshot": snapshot_version,
        "ops": [],
        "lastModified": datetime.utcnow().isoformat() + "Z",
    }
    write_changelog(workspace_path, new_changelog)

    print(f"[SchemaX] Created snapshot {snapshot_version}: {name}")
    print(f"[SchemaX] Snapshot file: {snapshot_metadata['file']}")
    print(f"[SchemaX] Ops included: {len(ops_with_ids)}")

    return project, snapshot_file


def get_uncommitted_ops_count(workspace_path: Path) -> int:
    """Get number of uncommitted operations"""
    changelog = read_changelog(workspace_path)
    return len(changelog["ops"])


def _calculate_state_hash(state: Any, ops_included: List[str]) -> str:
    """Calculate hash of state for integrity checking"""
    content = json.dumps({"state": state, "opsIncluded": ops_included}, sort_keys=True)
    return hashlib.sha256(content.encode()).hexdigest()


def _get_next_version(current_version: Optional[str], settings: Dict[str, Any]) -> str:
    """Get next version number"""
    if not current_version:
        return settings["versionPrefix"] + "0.1.0"

    # Parse version (e.g., "v0.1.0" or "0.1.0")
    import re

    match = re.search(r"(\d+)\.(\d+)\.(\d+)", current_version)
    if not match:
        return settings["versionPrefix"] + "0.1.0"

    major, minor, patch = match.groups()
    next_minor = int(minor) + 1

    return f"{settings['versionPrefix']}{major}.{next_minor}.0"

File: packages/vscode-extension/src/storage-v3.ts
/**
 * Storage Layer V3 - Provider-Aware
 * 
 * New storage layer that supports multiple catalog providers through the provider system.
 * Migrates from v2 storage format to v3 format with provider metadata.
 */

import * as vscode from 'vscode';
import * as fs from 'fs/promises';
import * as path from 'path';
import * as crypto from 'crypto';
import { v4 as uuidv4 } from 'uuid';
import { ProviderRegistry } from './providers/registry';
import { Provider } from './providers/base/provider';
import { Operation } from './providers/base/operations';
import { ProviderState } from './providers/base/models';

const SCHEMAX_DIR = '.schemax';
const PROJECT_FILENAME = 'project.json';
const CHANGELOG_FILENAME = 'changelog.json';
const SNAPSHOTS_DIR = 'snapshots';

// Project File V3 Types
interface ProviderConfig {
  type: string; // 'unity', 'hive', 'postgres'
  version: string; // Provider schema version
  config?: Record<string, any>; // Provider-specific config
}

interface ProjectSettings {
  autoIncrementVersion: boolean;
  versionPrefix: string;
  requireSnapshotForProd: boolean;
  allowDrift: boolean;
  requireComments: boolean;
  warnOnBreakingChanges: boolean;
}

interface SnapshotMetadata {
  id: string;
  version: string;
  name: string;
  ts: string;
  createdBy?: string;
  file: string;
  previousSnapshot: string | null;
  opsCount: number;
  hash: string;
  tags: string[];
  comment?: string;
}

interface Deployment {
  id: string;
  environment: string;
  ts: string;
  deployedBy?: string;
  snapshotId: string | null;
  opsApplied: string[];
  schemaVersion: string;
  sqlGenerated?: string;
  status: 'success' | 'failed' | 'rolled_back';
  error?: string;
  driftDetected: boolean;
  driftDetails?: any[];
}

export interface ProjectFileV3 {
  version: 3;
  name: string;
  provider: ProviderConfig;
  environments: string[];
  snapshots: SnapshotMetadata[];
  deployments: Deployment[];
  settings: ProjectSettings;
  latestSnapshot: string | null;
}

export interface ChangelogFile {
  version: 1;
  sinceSnapshot: string | null;
  ops: Operation[];
  lastModified: string;
}

export interface SnapshotFile {
  id: string;
  version: string;
  name: string;
  ts: string;
  createdBy?: string;
  state: ProviderState;
  opsIncluded: string[];
  previousSnapshot: string | null;
  hash: string;
  tags: string[];
  comment?: string;
}

/**
 * Get the project file path
 */
function getProjectFilePath(workspaceUri: vscode.Uri): string {
  return path.join(getSchemaxDir(workspaceUri), PROJECT_FILENAME);
}

/**
 * Get the changelog file path
 */
function getChangelogFilePath(workspaceUri: vscode.Uri): string {
  return path.join(getSchemaxDir(workspaceUri), CHANGELOG_FILENAME);
}

/**
 * Get the .schemax directory path
 */
function getSchemaxDir(workspaceUri: vscode.Uri): string {
  return path.join(workspaceUri.fsPath, SCHEMAX_DIR);
}

/**
 * Get the snapshots directory path
 */
function getSnapshotsDir(workspaceUri: vscode.Uri): string {
  return path.join(getSchemaxDir(workspaceUri), SNAPSHOTS_DIR);
}

/**
 * Get snapshot file path
 */
function getSnapshotFilePath(workspaceUri: vscode.Uri, version: string): string {
  return path.join(getSnapshotsDir(workspaceUri), `${version}.json`);
}

/**
 * Ensure .schemax/snapshots/ directory exists
 */
async function ensureSchemaxDir(workspaceUri: vscode.Uri): Promise<void> {
  const schemaxDir = getSchemaxDir(workspaceUri);
  const snapshotsDir = getSnapshotsDir(workspaceUri);
  
  try {
    await fs.mkdir(schemaxDir, { recursive: true });
    await fs.mkdir(snapshotsDir, { recursive: true });
  } catch (error) {
    // Directory might already exist, that's ok
  }
}

/**
 * Initialize a new project with provider selection
 */
export async function ensureProjectFile(
  workspaceUri: vscode.Uri,
  providerId: string = 'unity'
): Promise<void> {
  const projectPath = getProjectFilePath(workspaceUri);
  const changelogPath = getChangelogFilePath(workspaceUri);
  
  try {
    await fs.access(projectPath);
    // Project exists - check if migration is needed
    const content = await fs.readFile(projectPath, 'utf8');
    const project = JSON.parse(content);
    
    if (project.version === 2) {
      // Migrate from v2 to v3
      await migrateV2ToV3(workspaceUri, project, providerId);
    }
  } catch {
    // Create new v3 project
    const workspaceName = path.basename(workspaceUri.fsPath);
    
    // Verify provider exists
    const provider = ProviderRegistry.get(providerId);
    if (!provider) {
      throw new Error(`Provider '${providerId}' not found. Available providers: ${ProviderRegistry.getAllIds().join(', ')}`);
    }
    
    const newProject: ProjectFileV3 = {
      version: 3,
      name: workspaceName,
      provider: {
        type: providerId,
        version: provider.info.version,
      },
      environments: ['dev', 'test', 'prod'],
      snapshots: [],
      deployments: [],
      settings: {
        autoIncrementVersion: true,
        versionPrefix: 'v',
        requireSnapshotForProd: true,
        allowDrift: false,
        requireComments: false,
        warnOnBreakingChanges: true,
      },
      latestSnapshot: null,
    };
    
    const newChangelog: ChangelogFile = {
      version: 1,
      sinceSnapshot: null,
      ops: [],
      lastModified: new Date().toISOString(),
    };
    
    await ensureSchemaxDir(workspaceUri);
    await fs.writeFile(projectPath, JSON.stringify(newProject, null, 2), 'utf8');
    await fs.writeFile(changelogPath, JSON.stringify(newChangelog, null, 2), 'utf8');
    
    console.log(`[SchemaX] Initialized new v3 project: ${workspaceName} with provider: ${provider.info.name}`);
  }
}

/**
 * Migrate v2 project to v3 format
 */
async function migrateV2ToV3(
  workspaceUri: vscode.Uri,
  v2Project: any,
  providerId: string = 'unity'
): Promise<void> {
  console.log(`[SchemaX] Migrating project from v2 to v3...`);
  
  const provider = ProviderRegistry.get(providerId);
  if (!provider) {
    throw new Error(`Provider '${providerId}' not found`);
  }
  
  const v3Project: ProjectFileV3 = {
    version: 3,
    name: v2Project.name,
    provider: {
      type: providerId,
      version: provider.info.version,
    },
    environments: v2Project.environments || ['dev', 'test', 'prod'],
    snapshots: v2Project.snapshots || [],
    deployments: v2Project.deployments || [],
    settings: v2Project.settings || {
      autoIncrementVersion: true,
      versionPrefix: 'v',
      requireSnapshotForProd: true,
      allowDrift: false,
      requireComments: false,
      warnOnBreakingChanges: true,
    },
    latestSnapshot: v2Project.latestSnapshot || null,
  };
  
  // Migrate operations in changelog to add provider prefix
  try {
    const changelog = await readChangelog(workspaceUri);
    const migratedOps = changelog.ops.map((op: Operation) => {
      // Add provider field if missing
      if (!op.provider) {
        op.provider = providerId;
      }
      // Add provider prefix to operation type if missing
      if (!op.op.includes('.')) {
        op.op = `${providerId}.${op.op}`;
      }
      return op;
    });
    
    await fs.writeFile(
      getChangelogFilePath(workspaceUri),
      JSON.stringify({ ...changelog, ops: migratedOps }, null, 2),
      'utf8'
    );
  } catch (error) {
    console.warn('[SchemaX] Could not migrate changelog operations:', error);
  }
  
  // Write migrated project file
  const projectPath = getProjectFilePath(workspaceUri);
  await fs.writeFile(projectPath, JSON.stringify(v3Project, null, 2), 'utf8');
  
  console.log(`[SchemaX] Migration complete: v2 → v3`);
}

/**
 * Read project file
 */
export async function readProject(workspaceUri: vscode.Uri): Promise<ProjectFileV3> {
  const filePath = getProjectFilePath(workspaceUri);
  const content = await fs.readFile(filePath, 'utf8');
  const parsed = JSON.parse(content);
  
  // Auto-migrate if needed
  if (parsed.version === 2) {
    await migrateV2ToV3(workspaceUri, parsed);
    return readProject(workspaceUri); // Re-read after migration
  }
  
  if (parsed.version !== 3) {
    throw new Error(`Unsupported project version: ${parsed.version}`);
  }
  
  return parsed as ProjectFileV3;
}

/**
 * Write project file
 */
async function writeProject(workspaceUri: vscode.Uri, project: ProjectFileV3): Promise<void> {
  const filePath = getProjectFilePath(workspaceUri);
  const content = JSON.stringify(project, null, 2);
  await fs.writeFile(filePath, content, 'utf8');
}

/**
 * Read changelog file
 */
export async function readChangelog(workspaceUri: vscode.Uri): Promise<ChangelogFile> {
  const filePath = getChangelogFilePath(workspaceUri);
  try {
    const content = await fs.readFile(filePath, 'utf8');
    const parsed = JSON.parse(content);
    return parsed as ChangelogFile;
  } catch {
    // Changelog doesn't exist, create empty one
    const changelog: ChangelogFile = {
      version: 1,
      sinceSnapshot: null,
      ops: [],
      lastModified: new Date().toISOString(),
    };
    await fs.writeFile(filePath, JSON.stringify(changelog, null, 2), 'utf8');
    return changelog;
  }
}

/**
 * Write changelog file
 */
async function writeChangelog(workspaceUri: vscode.Uri, changelog: ChangelogFile): Promise<void> {
  const filePath = getChangelogFilePath(workspaceUri);
  changelog.lastModified = new Date().toISOString();
  const content = JSON.stringify(changelog, null, 2);
  await fs.writeFile(filePath, content, 'utf8');
}

/**
 * Read a snapshot file
 */
export async function readSnapshot(workspaceUri: vscode.Uri, version: string): Promise<SnapshotFile> {
  const filePath = getSnapshotFilePath(workspaceUri, version);
  const content = await fs.readFile(filePath, 'utf8');
  const parsed = JSON.parse(content);
  return parsed as SnapshotFile;
}

/**
 * Write a snapshot file
 */
async function writeSnapshot(workspaceUri: vscode.Uri, snapshot: SnapshotFile): Promise<void> {
  await ensureSchemaxDir(workspaceUri);
  const filePath = getSnapshotFilePath(workspaceUri, snapshot.version);
  const content = JSON.stringify(snapshot, null, 2);
  await fs.writeFile(filePath, content, 'utf8');
}

/**
 * Load current state using provider
 */
export async function loadCurrentState(workspaceUri: vscode.Uri): Promise<{
  state: ProviderState;
  changelog: ChangelogFile;
  provider: Provider;
}> {
  const project = await readProject(workspaceUri);
  const changelog = await readChangelog(workspaceUri);
  
  // Get provider
  const provider = ProviderRegistry.get(project.provider.type);
  if (!provider) {
    throw new Error(`Provider '${project.provider.type}' not found. Please ensure the provider is installed.`);
  }
  
  let state: ProviderState;
  
  if (project.latestSnapshot) {
    // Load latest snapshot
    const snapshot = await readSnapshot(workspaceUri, project.latestSnapshot);
    state = snapshot.state;
  } else {
    // No snapshots yet, start with empty state
    state = provider.createInitialState();
  }
  
  // Apply changelog ops using provider's state reducer
  state = provider.applyOperations(state, changelog.ops);
  
  return { state, changelog, provider };
}

/**
 * Append operations to changelog
 */
export async function appendOps(workspaceUri: vscode.Uri, ops: Operation[]): Promise<void> {
  const project = await readProject(workspaceUri);
  const changelog = await readChangelog(workspaceUri);
  const provider = ProviderRegistry.get(project.provider.type);
  
  if (!provider) {
    throw new Error(`Provider '${project.provider.type}' not found`);
  }
  
  // Validate operations
  for (const op of ops) {
    const validation = provider.validateOperation(op);
    if (!validation.valid) {
      const errors = validation.errors.map(e => `${e.field}: ${e.message}`).join(', ');
      throw new Error(`Invalid operation: ${errors}`);
    }
  }
  
  // Append ops
  changelog.ops.push(...ops);
  
  // Write back
  await writeChangelog(workspaceUri, changelog);
}

/**
 * Create a snapshot
 */
export async function createSnapshot(
  workspaceUri: vscode.Uri,
  name: string,
  version?: string,
  comment?: string,
  tags: string[] = []
): Promise<{ project: ProjectFileV3; snapshot: SnapshotFile }> {
  const project = await readProject(workspaceUri);
  const changelog = await readChangelog(workspaceUri);
  
  // Load current state
  const { state } = await loadCurrentState(workspaceUri);
  
  // Determine version
  const snapshotVersion = version || getNextVersion(project.latestSnapshot, project.settings);
  
  // Generate IDs for ops that don't have them (backwards compatibility)
  const opsWithIds = changelog.ops.map((op, index) => {
    if (!op.id) {
      op.id = `op_${index}_${op.ts}_${op.target}`;
    }
    return op;
  });
  
  // Calculate hash
  const hash = calculateStateHash(state, opsWithIds.map(op => op.id!));
  
  // Create snapshot file
  const snapshotFile: SnapshotFile = {
    id: `snap_${uuidv4()}`,
    version: snapshotVersion,
    name,
    ts: new Date().toISOString(),
    createdBy: process.env.USER || process.env.USERNAME || 'unknown',
    state,
    opsIncluded: opsWithIds.map(op => op.id!),
    previousSnapshot: project.latestSnapshot,
    hash,
    tags,
    comment,
  };
  
  // Write snapshot file
  await writeSnapshot(workspaceUri, snapshotFile);
  
  // Create snapshot metadata
  const snapshotMetadata: SnapshotMetadata = {
    id: snapshotFile.id,
    version: snapshotVersion,
    name,
    ts: snapshotFile.ts,
    createdBy: snapshotFile.createdBy,
    file: `.schemax/snapshots/${snapshotVersion}.json`,
    previousSnapshot: project.latestSnapshot,
    opsCount: opsWithIds.length,
    hash,
    tags,
    comment,
  };
  
  // Update project
  project.snapshots.push(snapshotMetadata);
  project.latestSnapshot = snapshotVersion;
  await writeProject(workspaceUri, project);
  
  // Clear changelog
  const newChangelog: ChangelogFile = {
    version: 1,
    sinceSnapshot: snapshotVersion,
    ops: [],
    lastModified: new Date().toISOString(),
  };
  await writeChangelog(workspaceUri, newChangelog);
  
  console.log(`[SchemaX] Created snapshot ${snapshotVersion}: ${name}`);
  console.log(`[SchemaX] Snapshot file: ${snapshotMetadata.file}`);
  console.log(`[SchemaX] Ops included: ${opsWithIds.length}`);
  
  return { project, snapshot: snapshotFile };
}

/**
 * Get uncommitted operations count
 */
export async function getUncommittedOpsCount(workspaceUri: vscode.Uri): Promise<number> {
  const changelog = await readChangelog(workspaceUri);
  return changelog.ops.length;
}

/**
 * Calculate hash of state for integrity checking
 */
function calculateStateHash(state: any, opsIncluded: string[]): string {
  const content = JSON.stringify({ state, opsIncluded });
  return crypto.createHash('sha256').update(content).digest('hex');
}

/**
 * Get next version number
 */
function getNextVersion(currentVersion: string | null, settings: ProjectSettings): string {
  if (!currentVersion) {
    return settings.versionPrefix + '0.1.0';
  }
  
  // Parse version (e.g., "v0.1.0" or "0.1.0")
  const versionMatch = currentVersion.match(/(\d+)\.(\d+)\.(\d+)/);
  if (!versionMatch) {
    return settings.versionPrefix + '0.1.0';
  }
  
  const [, major, minor, patch] = versionMatch;
  const nextMinor = parseInt(minor) + 1;
  
  return `${settings.versionPrefix}${major}.${nextMinor}.0`;
}

